### ISSUE TITLE
Accept and pass additional params to third party authorisation url

### ISSUE THREAD

**Issue by sbauch** [supertokens/supertokens-node#94]
I'm hoping to add a little bit of functionality to 3rd party oauth providers that allows a developer to get a value from user input and include that as a query param in the service's authorisation url.

NextAuth.js handles this well IMO, and [provides some examples](https://next-auth.js.org/getting-started/client#additional-params) for when this would be useful and further points at an [OIDC spec for more examples](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest).

A further example that better illustrates needing user input is Google OAuth's `hd` param, which restricts Google users to using accounts from the domain provided to the `hd` param in the authorization URL.

I maintain an open source OAuth service [Osso](https://github.com/enterprise-oss/osso) that wraps SAML auth in an OAuth flow. SAML authentication is instance based - you must send a user to the expected IDP instance, and so you need to determine who the user is in order to route them to the right IDP.

So we want to be able to craft a login form that asks for the user's email and includes it as a param to the third party authorisation URL.

It seems like this URL construction is happening here - https://github.com/supertokens/supertokens-node/blob/master/lib/ts/recipe/thirdparty/api/authorisationUrl.ts#L68-L70

A naive approach might be to append the other params that are included in the request. I suppose there is some danger in including wayward params however.

@rishabhpoddar suggested on discord that the params type be extended to accept a function. 

I think the suggestion is to expand this type definition to accept functions as values - https://github.com/supertokens/supertokens-node/blob/master/lib/ts/recipe/thirdparty/types.ts#L35

I might suggest keeping that as a `{ [key: string]: string }` typed object, and instead add an additional key `dynamicParams`.

So something like this?

```ts
export type TypeProviderGetResponse = {
    accessTokenAPI: {
        url: string;
        params: { [key: string]: string }; // Will be merged with our object
    };
    authorisationRedirect: {
        url: string;
        params: { [key: string]: string };
        dynamicParams: { [key: string]: (request: Request) => string; } 
    };
    getProfileInfo: (authCodeResponse: any) => Promise<UserInfo>;
};
```

Then, if we were to update the third party strategies to pass these along - 

```ts
let dynamicParamsString = Object.entries(providerInfo.authorisationRedirect.dynamicParams).reduce(
  (dynamicParams, [key, getter]) => ({
    ...dynamicParams,
    [key]: getter(request),
  }
), {}).toString()

let url = `${providerInfo.authorisationRedirect.url}?${paramsString}&${dynamicParamsString}`;

```

And then in userland when using google's `hd` param or some locale param that i made up:

```ts
ThirdParty.Google({
  clientSecret: "GOOGLE_CLIENT_SECRET",
  clientId: "GOOGLE_CLIENT_ID",
  authorisationUrl: {
    dynamicParams: {
      hd: (request: Request) => request.query.domain,
      locale: (request: Request) => request.acceptsLanguages[0]
    }
  }
})
```

I think there are a couple other places where types would need to be updated, but does this feel like a reasonable addition?

**Comment by rishabhpoddar** [comment]
This seems like a good approach, however, I would prefer if the types are updated to:
```ts
params: { [key: string]: string | [key: string]: (request: Request) => string };
```

This way, in userland, we can do:
```ts
ThirdParty.Google({
  clientSecret: "GOOGLE_CLIENT_SECRET",
  clientId: "GOOGLE_CLIENT_ID",
  authorisationUrl: {
    params: {
      hd: (request: Request) => request.query.domain,
      locale: (request: Request) => request.acceptsLanguages[0],
      someOtherKey: "someCOnstant"
    }
  }
})
```

Please feel free to issue a PR for this (base as branch 4.0).

**Comment by sbauch** [comment]
it certainly works!

here's 2 reasons I proposed my approach - 

1. generally, expanding the type to accept a string or a function kinda starts to move away from anything actually being typed
2. when constructing the authorisation url, for each `param`, you'd have to now check it's type to determine whether it's a callable function or a string

That said it's your project, I'm happy to take your suggestion 😁

**Comment by rishabhpoddar** [comment]
Thanks for your reply. I appreciate your counter arguments and agree with them too. 

However, the only reason I still prefer expanding the type is because it's slightly more easy from a documentation perspective (I feel it will require lesser explanation), and optimising for easier docs is always our number one priority when deciding on issues like these.

Would you be interested in creating a PR for this?

**Comment by sbauch** [comment]
Sure that's fair.

I'd be happy to create a PR, but probably won't get to it for a few days.

**Comment by rishabhpoddar** [comment]
@sbauch I have released a new node version (v4.0.1) with the above change! Thanks for the PR once again. Have a great day
